src/platforms/emscripten/src/lib/platform_nifs.c

```diff
@@ -157,6 +158,289 @@ static term nif_emscripten_promise_reject(Context *ctx, int argc, term argv[])
+{
+    UNUSED(argc);
+    term script_term = argv[0];
+
+    int ok;
+    char *script = interop_term_to_string(script_term, &ok);
+    if (UNLIKELY(!ok)) {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+    // Trap caller waiting for completion
+    context_update_flags(ctx, ~NoFlags, Trap);
+    // script will be freed as it's passed as satellite
+    emscripten_dispatch_to_thread(emscripten_main_runtime_thread_id(), EM_FUNC_SIG_VIII, do_run_script_tracked, script, script, ctx->process_id, ctx->global);
+    return term_invalid_term();
+}
+
+// clang-format off
+EM_JS(char *, js_get_tracked_objects, (uint32_t *keys_ptr, uint32_t keys_n, uint32_t **sizes, uint8_t **statuses, uint32_t *objects_n, uint32_t *strings_n, uint32_t *all_byte_size), {
+    const OK = 0;
+    const BAD_KEY = 1;
+    const NOT_STRING = 2;
+
+    const keysOffset = keys_ptr / HEAPU32.BYTES_PER_ELEMENT;
+    const keys = [...HEAPU32.subarray(keysOffset, keysOffset + keys_n)];
+
+    const objects = Module['onGetTrackedObjects'](keys);
+        if (!Array.isArray(objects)) {
+        return 0;
+    }
+    const n = objects.length;
+
+    if (n === 0) {
+        return 0;
+    }
+
+    const sizesPtr = Module['_malloc'](n * HEAPU32.BYTES_PER_ELEMENT);
+    const statusPtr = Module['_malloc'](n * HEAPU8.BYTES_PER_ELEMENT);
+    let allByteSize = 0;
+    let stringsN = 0;
+
+    for (let i = 0; i < n; ++i) {
+        let status = OK;
+        let byteSize = 0;
+        const object = objects[i];
+                if (object === undefined) {
+            status = BAD_KEY;
+        } else if (typeof object !== "string") {
+            status = NOT_STRING;
+        } else {
+            byteSize = lengthBytesUTF8(object);
+            stringsN += 1;
+        }
+        HEAPU8[statusPtr / HEAPU8.BYTES_PER_ELEMENT + i] = status;
+        HEAPU32[sizesPtr / HEAPU32.BYTES_PER_ELEMENT + i] = byteSize;
+        allByteSize += byteSize;
+    }
+
+    const stringsPtr = Module['_malloc'](allByteSize * HEAPU8.BYTES_PER_ELEMENT);
+    let currentStringsPtr = stringsPtr;
+    for (let i = 0; i < n; ++i) {
+        const status = HEAPU8[statusPtr / HEAPU8.BYTES_PER_ELEMENT + i];
+        if (status !== OK) {
+            continue;
+        }
+        const string = objects[i];
+        const size = HEAPU32[sizesPtr / HEAPU32.BYTES_PER_ELEMENT + i];
+        // stringToUTF8 includes null byte which we don't need
+        stringToUTF8(string, currentStringsPtr, size+1);
+        currentStringsPtr += size;
+    }
+
+    HEAPU32[statuses / HEAPU32.BYTES_PER_ELEMENT] = statusPtr;
+    HEAPU32[sizes / HEAPU32.BYTES_PER_ELEMENT] = sizesPtr;
+    HEAPU32[objects_n / HEAPU32.BYTES_PER_ELEMENT] = n;
+    HEAPU32[strings_n / HEAPU32.BYTES_PER_ELEMENT] = stringsN;
+    HEAPU32[all_byte_size / HEAPU32.BYTES_PER_ELEMENT] = allByteSize;
+
+    return stringsPtr;
+});
+// clang-format on
+
+static void do_get_tracked_objects(uint32_t *ref_keys, size_t keys_n, int32_t sync_caller_pid, GlobalContext *global)
```

Comment:
Right now we are using `uint32_t` for process ids. Also we are using the extended `process_id` instead of `pid` when talking about the id itself not encoded as a term pid.